src/core/Component.ts

```diff
@@ -1,16 +1,79 @@
 import { ComponentType } from 'src/components'
 import { uuid } from 'src/utils'
-import { AbstractDataObject, DataObject } from './DataObject'
 
-export abstract class Component extends AbstractDataObject {
+export type ComponentAttributes = Record<string, string | number | { serialize(): string }>
+
+export abstract class Component<T extends ComponentAttributes> {
   // @internal
   uuid: string = uuid()
 
   // @internal
   readonly type: ComponentType
 
-  constructor(type: ComponentType) {
-    super()
+  // @internal
+  protected readonly _attributes: T = {} as any
+
+  // @internal
+  private _isDirty: boolean = true
+
+  // @internal
+  private cache: Record<string, string> = {}
+
+  attributes: T
+
+  constructor(type: ComponentType, attributes: T) {
     this.type = type
+    this._attributes = attributes
+    this.attributes = new Proxy(this._attributes, {
+      get(target: T, prop: string, c) {
+        return target[prop]
+      },
+      set(target: T, prop: string, value: any) {
+        target[prop] = value
+        this.isDirty = true
+        return true
+      }
+    })
+  }
+
+  set isDirty(flag: boolean) {
+    // TODO (telemetry): track dirty flags
+    this._isDirty = flag
+  }
+
+  get isDirty() {
+    // Dirty until set otherwise
+    if (this._isDirty) return true
+
+    // Lazy: check nested structures
+    for (let key in this._attributes) {
+      const field = this._attributes[key]
+      const fieldType = typeof field
+
+      // Scalar types will mark the component as dirty when individually set
+      if (fieldType !== 'number' && fieldType !== 'string' && field !== null) {
+        if ('serialize' in field) {
+          // TODO: are strings the best serialization solution?
+          const val = (field as any).serialize()
```

Comment:
We can move `{ serialize(): string  }` into `HashableObject`, rename serialize to `getHash` and use that instead of any.

Note: if we change it one day this won't save us because we are forcing the type. We still rely on `ComponentAttributes` and I think that's enough.
Note 2: `getHash()` should be a lightweight string representation of the object: maybe instead of `Vector(1,1,1).toString()` (json-like) it can return `"111"` (for `Vector(1,1,1)`)

Reply:
Oops, imagine vector2 :P

Reply:
> We can move { serialize(): string } into HashableObject, rename serialize to getHash and use that instead of any.

I think that's a good middleground, but not super necessary. If you think it adds some value I am in favour

Reply:
check the last commit ;)